File: src/morie/fn/aitsbc.py
# morie.fn -- function file (rootcoder007/morie)
"""Subcomposition: closure of a selected subset of parts."""

from . import _tail1core as C

from ._richresult import RichResult

__all__ = ["subcomp", "aitchison_subcomposition"]


def subcomp(x, parts, total=1.0):
    """Form the subcomposition on a selected subset of parts.

    Subcompositional coherence -- that an analysis of a subset gives
    the same answer whether or not the other parts were ever measured
    -- is the property Aitchison uses to rule out correlations of raw
    proportions.  The subcomposition is the operation that property is
    stated about.

    ``parts`` is a list of ONE-BASED part indices, matching the R arm.

    Formula: sub(x; S) = C( x_i : i in S )

    Parameters
    ----------
    x : array-like
        Strictly positive vector of parts.
    parts : sequence of int
        One-based indices of the parts retained (at least two).
    total : float
        Constant the subcomposition sums to.

    Returns
    -------
    RichResult
        ``composition``, ``parts``, ``total``, ``D_sub``, ``D``.

    References
    ----------
    Aitchison (1986), The Statistical Analysis of Compositional Data,
    Chapter 2, where the subcomposition is defined as the closure of
    the selected subvector.
    """
    x = C.vec(x)
    if any(v <= 0 for v in x):
        raise ValueError("compositions must be strictly positive")
    D = len(x)
    idx = [int(i) for i in parts]
    if len(idx) < 2:
        raise ValueError("a subcomposition needs at least two parts")
    if any(not 1 <= i <= D for i in idx):
        raise ValueError("parts must be one-based indices in 1..D")
    sub = [x[i - 1] for i in idx]
    s = sum(sub)
    k = float(total)
    return RichResult(payload={
        "composition": [k * v / s for v in sub], "parts": idx, "total": k,
        "D_sub": len(idx), "D": D, "method": "Subcomposition"})


aitchison_subcomposition = subcomp


def cheatsheet():
    return "aitsbc: sub(x; S) = C(x_S)"

File: r-package/morie/R/aitsbc.R
# SPDX-License-Identifier: AGPL-3.0-or-later
#' Subcomposition formed from a selected set of parts.
#'
#' Formula: sub(x; S) = C( x_i : i in S )
#'
#' @param x Composition with strictly positive parts.
#' @param parts 1-based indices of the parts to keep; at least two, no repeats.
#' @param total Constant kappa the closure sums to.
#'
#' @return List with ``composition``, ``parts``, ``total``, ``D``, ``D_full``.
#' @references Aitchison, J. (1986), The Statistical Analysis of Compositional Data, Chapman and Hall, is this shelf's primary book and is NOT in the reference library, so it could not be read; see EXTERNAL_SOURCES.md.  A subcomposition is the closure of a selected subset of parts.  Log-ratios between retained parts are unchanged by the operation, which is the property that makes subcompositional coherence a requirement on compositional methods.  Indices are 1-based in BOTH language arms.  Implemented in the standard published form.  The log-ratio algebra it rests on was verified against Mateu-Figueras, Pawlowsky-Glahn and Egozcue, arXiv:0802.2643 Sect. 4.1 (fetched and archived), but this particular definition is not printed there and could not be checked against Aitchison's own text.
#' @export
Compsubcomp <- function(x, parts, total = 1) {
  x <- .t1_vec(x); D0 <- length(x)
  if (any(x <= 0)) stop("compositions must be strictly positive")
  idx <- as.integer(parts)
  if (length(idx) < 2L) stop("a subcomposition needs at least two parts")
  if (anyDuplicated(idx)) stop("parts must not repeat")
  if (any(idx < 1L | idx > D0)) stop("parts must be 1-based indices into x")
  sub <- x[idx]; s <- sum(sub); k <- as.numeric(total)
  .t1_result(composition = k * sub / s, parts = idx, total = k,
             D = length(idx), D_full = D0, method = "Subcomposition")
}
